p2p/net/nat/internal/nat/upnp.go

```diff
@@ -16,197 +16,115 @@ import (
-					client := &internetgateway2.WANIPConnection2{ServiceClient: goupnp.ServiceClient{
-						SOAPClient: srv.NewSOAPClient(),
-						RootDevice: dev.Root,
-						Service:    srv,
-					}}
-					_, isNat, err := client.GetNATRSIPStatusCtx(ctx)
-					if err == nil && isNat {
-						select {
-						case res <- &upnp_NAT{client, make(map[int]int), "UPNP (IG2-IP2)", dev.Root}:
-						case <-ctx.Done():
-						}
-					}
-
-				case internetgateway2.URN_WANPPPConnection_1:
-					client := &internetgateway2.WANPPPConnection1{ServiceClient: goupnp.ServiceClient{
-						SOAPClient: srv.NewSOAPClient(),
-						RootDevice: dev.Root,
-						Service:    srv,
-					}}
-					_, isNat, err := client.GetNATRSIPStatusCtx(ctx)
-					if err == nil && isNat {
-						select {
-						case res <- &upnp_NAT{client, make(map[int]int), "UPNP (IG2-PPP1)", dev.Root}:
-						case <-ctx.Done():
-						}
-					}
+func discoverSearchTarget(ctx context.Context, target string) (nats []NAT, errs []error) {
+	// find devices
+	devs, err := goupnp.DiscoverDevicesCtx(ctx, target)
+	if err != nil {
+		errs = append(errs, err)
+		return
+	}
 
-				}
-			})
+	for _, dev := range devs {
+		if dev.Err != nil {
+			errs = append(errs, dev.Err)
+			continue
 		}
-
-	}()
-	return res
+		dev.Root.Device.VisitServices(serviceVisitor(ctx, dev.Root, &nats, &errs))
+	}
+	return
 }
 
-func discoverUPNP_GenIGDev(ctx context.Context) <-chan NAT {
-	res := make(chan NAT, 1)
-	go func() {
-		defer close(res)
+// discoverUPNP_GenIGDev is a fallback for routers that fail to respond to our
+// targetted SSDP queries. It will query all devices and try to find any
+// InternetGatewayDevice.
+func discoverUPNP_GenIGDev(ctx context.Context) (nats []NAT, errs []error) {
+	DeviceList, err := ssdp.Search(ssdp.All, 5, "")
+	if err != nil {
+		errs = append(errs, err)
+		return
+	}
 
-		DeviceList, err := ssdp.Search(ssdp.All, 5, "")
-		if err != nil {
-			return
+	const maxIGDevs = 3
+	foundIGDevs := 0
+	for _, Service := range DeviceList {
+		if !strings.Contains(Service.Type, "InternetGatewayDevice") {
+			continue
 		}
-		var gw ssdp.Service
-		for _, Service := range DeviceList {
-			if strings.Contains(Service.Type, "InternetGatewayDevice") {
-				gw = Service
-				break
-			}
+		if foundIGDevs >= maxIGDevs {
+			break
 		}
+		foundIGDevs++
 
-		DeviceURL, err := url.Parse(gw.Location)
+		DeviceURL, err := url.Parse(Service.Location)
 		if err != nil {
-			return
+			errs = append(errs, err)
+			continue
 		}
 		RootDevice, err := goupnp.DeviceByURLCtx(ctx, DeviceURL)
 		if err != nil {
-			return
+			errs = append(errs, err)
+			continue
 		}
 
-		RootDevice.Device.VisitServices(func(srv *goupnp.Service) {
-			if ctx.Err() != nil {
-				return
+		RootDevice.Device.VisitServices(serviceVisitor(ctx, RootDevice, &nats, &errs))
+	}
+	return
+}
+
+func serviceVisitor(ctx context.Context, rootDevice *goupnp.RootDevice, outNats *[]NAT, outErrs *[]error) func(srv *goupnp.Service) {
```

Comment:
We need a comment somewhere (e.g. on this function or the top level caller of all the upnp discovery mechanisms) describing why we'll handle whatever service type gets sent to us even if we searched for a totally different one (e.g. searching for IG1 and getting IP2 which comes from IG2). This will help us later if there's a problem.

Note: do we want / need this behavior anymore now that we're doing the sequential lookups to deal with badly implemented routers (e.g. fritzbox)?

Reply:
This will now collect an error if it sees a v2 service on a v1 device.